File: examples/calibration/evolution/sources/granite.ts
type Entry = { value: string };

class Store {
  present = true;
  failWrite = false;
  failSync = false;

  remove(_oldKey: string): void {
    this.present = false;
  }

  write(_next: Entry): void {
    if (this.failWrite) {
      throw new Error("write");
    }
    this.present = true;
  }

  sync(): void {
    if (this.failSync) {
      throw new Error("sync");
    }
  }
}

export function replaceEntry(store: Store, oldKey: string, next: Entry): void {
  store.remove(oldKey);
  store.write(next);
  store.sync();
}

File: examples/calibration/evolution/sources/granite.rs
enum Error {
    Failed,
}

struct Entry;

struct Store {
    present: bool,
    fail_write: bool,
    fail_sync: bool,
}

impl Store {
    fn remove(&mut self, _: &str) -> Result<(), Error> {
        self.present = false;
        Ok(())
    }

    fn write(&mut self, _: &Entry) -> Result<(), Error> {
        if self.fail_write {
            return Err(Error::Failed);
        }
        self.present = true;
        Ok(())
    }

    fn sync(&self) -> Result<(), Error> {
        if self.fail_sync {
            return Err(Error::Failed);
        }
        Ok(())
    }
}

fn replace_entry(store: &mut Store, old_key: &str, next: &Entry) -> Result<(), Error> {
    store.remove(old_key)?;
    store.write(next)?;
    store.sync()?;
    Ok(())
}
